File: Ajeong-Im/8주차/BJ_16724_피리_부는_사나이.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);
const map = input.slice(1).map((line) => line.split(""));

// 상태: 0 = 미방문, 1 = 탐색중, 2 = 탐색완료
const visited = Array.from({ length: N }, () => Array(M).fill(0));

const direction = { U: [-1, 0], D: [1, 0], L: [0, -1], R: [0, 1] };

let answer = 0;

function dfs(x, y) {
  visited[x][y] = 1; // 탐색 중

  const [dx, dy] = direction[map[x][y]];
  const nx = x + dx;
  const ny = y + dy;

  if (visited[nx][ny] === 0) {
    dfs(nx, ny);
  } else if (visited[nx][ny] === 1) {
    answer++;
  }

  visited[x][y] = 2; // 탐색 완료
}

for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    if (visited[i][j] === 0) {
      dfs(i, j);
    }
  }
}

console.log(answer);

File: youlimsongs/8주차/BJ_16724_피리 부는 사나이.py
#그래프 탐색하면서 아래 옆 위로 이동하는 함수
#한 지점에서 잡으면서 그래프 탐색 -> 안전존 1개
#방문안한 지점 잡아서 또 탐색 -> 안전존 1개

n,m = map(int, input().split())
graph = []
for _ in range(n):
    graph.append(list(input()))

visited = [[False]*m for _ in range(n)]
cnt = 0

def check_dir(i,j):
    match graph[i][j]:
        case 'U':
            return (i-1, j)
        case 'D':
            return (i+1, j)
        case 'R':
            return (i, j+1)
        case 'L':
            return (i, j-1)

def dfs(y, x):
    visited[y][x] = 1
    ny, nx = check_dir(y, x)

    if visited[ny][nx] == 0:
        dfs(ny, nx)
    elif visited[ny][nx] == 1:  #사이클 완성
        global cnt
        cnt += 1

    visited[y][x] = 2  # 방문 마침

for i in range(n):
    for j in range(m):
        if visited[i][j] == 0:
            dfs(i,j)

print(cnt)
